File: packages/dataset_config_dart/lib/src/models/tag_filter.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'tag_filter.freezed.dart';
part 'tag_filter.g.dart';

/// Tag-based filter for including/excluding items by their tags.
@freezed
sealed class TagFilter with _$TagFilter {
  const factory TagFilter({
    @JsonKey(name: 'include_tags') List<String>? includeTags,
    @JsonKey(name: 'exclude_tags') List<String>? excludeTags,
  }) = _TagFilter;

  factory TagFilter.fromJson(Map<String, dynamic> json) =>
      _$TagFilterFromJson(json);
}

/// Check whether a set of [itemTags] matches the given [filter].
///
/// Returns `true` if:
/// - All include_tags (if any) are present in [itemTags]
/// - No exclude_tags (if any) are present in [itemTags]
bool matchesTagFilter(List<String> itemTags, TagFilter filter) {
  if (filter.includeTags != null &&
      filter.includeTags!.isNotEmpty &&
      !filter.includeTags!.every((t) => itemTags.contains(t))) {
    return false;
  }
  if (filter.excludeTags != null &&
      filter.excludeTags!.isNotEmpty &&
      filter.excludeTags!.any((t) => itemTags.contains(t))) {
    return false;
  }
  return true;
}

File: packages/dataset_config_python/src/dataset_config_python/models/tag_filter.py
"""Tag-based filter for including/excluding items by their tags."""

from __future__ import annotations

from pydantic import BaseModel


class TagFilter(BaseModel):
    """Tag-based filter for including/excluding items."""

    include_tags: list[str] | None = None
    exclude_tags: list[str] | None = None


def matches_tag_filter(item_tags: list[str], tag_filter: TagFilter) -> bool:
    """Check whether a set of item_tags matches the given filter.

    Returns True if:
    - All include_tags (if any) are present in item_tags
    - No exclude_tags (if any) are present in item_tags
    """
    if tag_filter.include_tags and not all(
        t in item_tags for t in tag_filter.include_tags
    ):
        return False
    if tag_filter.exclude_tags and any(
        t in item_tags for t in tag_filter.exclude_tags
    ):
        return False
    return True
